object_store/src/aws/mod.rs

```
@@ -1002,76 +1017,74 @@ impl AmazonS3Builder {
         let checksum = self.checksum_algorithm.map(|x| x.get()).transpose()?;
 
-        let credentials = match (self.access_key_id, self.secret_access_key, self.token) {
-            (Some(key_id), Some(secret_key), token) => {
-                info!("Using Static credential provider");
-                let credential = AwsCredential {
-                    key_id,
-                    secret_key,
-                    token,
-                };
-                Arc::new(StaticCredentialProvider::new(credential)) as _
-            }
-            (None, Some(_), _) => return Err(Error::MissingAccessKeyId.into()),
-            (Some(_), None, _) => return Err(Error::MissingSecretAccessKey.into()),
-            // TODO: Replace with `AmazonS3Builder::credentials_from_env`
-            _ => match (
-                std::env::var("AWS_WEB_IDENTITY_TOKEN_FILE"),
-                std::env::var("AWS_ROLE_ARN"),
-            ) {
-                (Ok(token_path), Ok(role_arn)) => {
-                    info!("Using WebIdentity credential provider");
-
-                    let session_name = std::env::var("AWS_ROLE_SESSION_NAME")
-                        .unwrap_or_else(|_| "WebIdentitySession".to_string());
-
-                    let endpoint = format!("https://sts.{region}.amazonaws.com");
-
-                    // Disallow non-HTTPs requests
-                    let client = self
-                        .client_options
-                        .clone()
-                        .with_allow_http(false)
-                        .client()?;
-
-                    let token = WebIdentityProvider {
-                        token_path,
-                        session_name,
-                        role_arn,
-                        endpoint,
-                    };
-
-                    Arc::new(TokenCredentialProvider::new(
+        let credentials = if let Some(credentials) = self.credentials {
+            credentials
```

I think it  is worth considering  a `warn!` here if `access_key_id` or `secret_access_key` is set that they are being ignored in favor of the other credentials. I know the docs say this is the behavior but I think as a user I would find it somewhat confusing if the access_key_id got silently ignored -- I would rather the library loudly complained so I can fix it

Maybe in that vein, the builder should return an error if conflicting credentials are supplied
 🤔

---

I'm not sure about this as some systems may source config from the environment, which may then contain redundant credentials. I've filed https://github.com/apache/arrow-rs/issues/4240 to track this, and will come back to this after I've had some time to think on it, and others have had time to weigh in